### PR TITLE
Add geodetic representations

### PR DESCRIPTION
**Pull request by astrojuanlu**
### Description

In line with discussions at #4261, this pull request aims to add geodetic representations.

I started with `WGS84GeodeticRepresentation` to touch base and get a few eyes on my approach. In particular, I don't think we can cleanly have a generic `GeodeticRepresentation` where the ellipsoid is an extra parameter, retained in roundtrip transformations, but I would be glad to be proven wrong.

The tests are almost a copy-paste of `TestSphericalRepresentation`, with a few changes:

- `height` instead of `distance`
- `height` can only be a `Quantity`, not a `Distance`
- I didn't add differentials (not sure if worth it)

Besides, I also adapted `test_gc2gd` and `test_gd2gc` from `test_earth.py` to make sure we get the same results.

### CODE REVIEW

File: astropy/coordinates/representation.py
Comment (mhvk) [inline astropy/coordinates/representation.py:2275]: While you are at it: do change this to `lon << u.radian`, etc., that avoids an extra copy.

**Review by mhvk — COMMENTED** [review]
@astrojuanlu - this looks good! I'm actually pleasantly surprised to see how relatively simple it is to add - the work is mostly in the tests!

Beyond the small in-line comments: I think these representations belong in `earth.py` as they are not useful as general representations. Similarly, I'd put the tests in `test_earth`. (Or perhaps a new `earth_representations` or even an `earth/` module. Though seems a bit overkill right now.)

I agree that for now we can ignore velocities - we can add the ability to mimic observations from airplanes when the need arises!

How were you thinking of proceeding? A first step might simply be to remove the transformation machinery in `EarthLocation` and instead go through these representations inside `from_geodetic` and `lon, lat, heigh` - `to_geodetic()` may need some kind of deprecation warning as right now it returns a `namedtuple` and people may count on it. (I think that is OK, we should move to `.represent_as` anyway, making it more frame-like.)

p.s. I'll change this to draft right now just to be sure we don't accidentally merge.

File: astropy/coordinates/representation.py
Comment (astrojuanlu) [reply astropy/coordinates/representation.py:2260]: Good point about the class attribute, I'll explore that!

About `erfa.gd2gc1` though: I can't seem to find it :question:

File: astropy/coordinates/representation.py
Comment (astrojuanlu) [reply astropy/coordinates/representation.py:2249]: Will do! I was dissuaded from doing it because I didn't want the parallax, cosmology, etc But it's good to enforce the unit here.

File: astropy/coordinates/representation.py
Comment (astrojuanlu) [reply astropy/coordinates/representation.py:2275]: Good catch!

**Comment by astrojuanlu** [comment]
Thanks for the prompt review @mhvk !

I agree with moving this to `earth.py` and with all your comments. I wanted to think about the way forward with `EarthLocation` today but I couldn't, so I don't have any ideas yet - will give it some thought tomorrow.

File: astropy/coordinates/representation.py
Comment (mhvk) [reply astropy/coordinates/representation.py:2260]: That was a type (1 instead of back tick), i.e., just use `erfa.gd2gc`. Sorry!

File: astropy/coordinates/representation.py
Comment (mhvk) [reply astropy/coordinates/representation.py:2249]: Indeed, maybe we should have a `Length` class or so... But my suggestion was indeed only to enforce the unit, which seems easiest through the class.

**Comment by mkbrewer** [comment]
> I agree that for now we can ignore velocities - we can add the ability to mimic observations from airplanes when the need arises!

Don't forget balloons!  https://askdrphilastro.wordpress.com/2019/12/15/blast-antarctica-2019-2020/

**Comment by mhvk** [comment]
> Don't forget balloons!

How can I forget my own colleague!  So, we will have to get there...

File: astropy/coordinates/representation.py
Comment (astrojuanlu) [reply astropy/coordinates/representation.py:2249]: Besides, notice that in other representation classes, `Distance` is not used in `attr_classes`, but in the initializer. See for example:

https://github.com/astropy/astropy/blob/0ff46cb5305d288af47bd3de4432c7dc941b33cb/astropy/coordinates/representation.py#L1829-L1848

File: astropy/coordinates/representation.py
Comment (astrojuanlu) [reply astropy/coordinates/representation.py]: Oh, got it!

File: astropy/coordinates/tests/test_earth.py
Comment (astrojuanlu) [inline astropy/coordinates/tests/test_earth.py:44]: These are copy-pasted from `test_representations.py`. Should we move them to `astropy.utils` or `astropy.tests.helper`, or is it not too bad to have them duplicated?

**Comment by astrojuanlu** [comment]
Addressed the first batch of comments, going to look into `EarthLocation` now.

File: astropy/coordinates/representation.py
Comment (astrojuanlu) [reply astropy/coordinates/representation.py]: ...which, interestingly, doesn't work: https://github.com/astropy/astropy/issues/11104

**Comment by astrojuanlu** [comment]
What I did in the end was to deprecate `.to_geodetic` and `.to_geocentric` without changing their return values, and added a new `.represent_as`.

This is ready for a second pass!

**Comment by astrojuanlu** [comment]
Many test failures due to the `AstropyDeprecationWarning` I introduced - I guess no internal calls to deprecated methods are allowed. Will fix it.

**Comment by mkbrewer** [comment]
I would just note that many locations on Earth have heights that are below that of the ellipsoid. I seems a bit unreasonable to force a different method for creating and `EarthLocation` in these instances.

**Comment by mhvk** [comment]
Darn, we'd need `allow_negative`. I think for now it would be best to just proceed with a `Quantity` - we can deal with insisting it is a length differently/later.

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:596]: Agree with the deprecation message, but in analogy with a frame or sky coord, I think there is little against still having the properties for particular representations, i.e., I think we should allow (and also refer to in the warning message),
```
el.{wgs84,...}geodetic
el.cartesian
```

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:608]: Again seeing `EarthLocation` as more like a frame/skycoord rather than a representation, I think we can allow strings. I'd suggest following the nomenclature in their `represent_as` method and calling it `base`.

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:875]: The classes are `Longitude` and `Latitude` - might as well be precise...

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:869]: I think this will get registered in `REPRESENTATION_CLASSES`, which is not ideal... Pinging @adrn for ideas; should we allow a `name = False` explicit way to avoid registration? (Note: this should not have to hold up this PR!)

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:927]: One disadvantage of setting this docstring is that you lose the nice information you added in the `Base...`.  For a possible solution, see https://docs.astropy.org/en/latest/api/astropy.utils.decorators.format_doc.html#astropy.utils.decorators.format_doc

File: astropy/coordinates/tests/test_earth.py
Comment (mhvk) [reply astropy/coordinates/tests/test_earth.py:44]: I'd just import then from `test_representation` for now (also done in `test_representation_methods`...)

File: astropy/coordinates/tests/test_earth.py
Comment (mhvk) [inline astropy/coordinates/tests/test_earth.py:442]: I would shrink this class substantially - many of the tests are done on `Spherical` to just have them done once, no need to do them again. Also, do re-align for the longer name...

File: astropy/coordinates/tests/test_earth.py
Comment (mhvk) [inline astropy/coordinates/tests/test_earth.py:619]: Probably good to put a TODO on top that we do not yet have differentials (and then remove the commented out code here)

File: astropy/coordinates/tests/test_earth.py
Comment (mhvk) [inline astropy/coordinates/tests/test_earth.py:672]: Missing from the tests is the new `represent_as` method.

File: astropy/coordinates/tests/test_earth.py
Comment (mhvk) [reply astropy/coordinates/tests/test_earth.py:442]: Thinking more, since all new tests are (very nicely) just focussed on the representations, how about a new `test_geodetic_representations`?

**Review by mhvk — COMMENTED** [review]
@astrojuanlu - this is starting to look very nice.  Bit of a mix of larger and probably overly detailed comments.

More general one is where we stop. One option is actually a bit *before* where you are now, i.e., not add deprecation messages yet, but just introduce the new representations and use them *internally* only for the transformations. This has the advantage that no tests will fail yet. Arguably, in that case, we would also not yet add the `represent_as` method (which is not tested yet, so that is fine, I think).

The other is perhaps a bit further than you are now, whether `EarthLocation` stops becoming a `Quantity` and instead stores the representation in `.data` like a frame (or perhaps as an intermediate suggestion still shares the data with a `CartesianRepresentation`).

But my sense would be to take the smaller step suggested above, so we can get this in and follow-up on it.

**Comment by astrojuanlu** [comment]
> But my sense would be to take the smaller step suggested above, so we can get this in and follow-up on it.

I agree - I will remove the deprecation warnings and remove `represent_as` for now, and follow up in a future PR. Addressing your comments now.

File: astropy/coordinates/earth.py
Comment (astrojuanlu) [reply astropy/coordinates/earth.py:596]: Removed the deprecation message for now, but saving this for the future.

File: astropy/coordinates/earth.py
Comment (astrojuanlu) [reply astropy/coordinates/earth.py:608]: Removed `represent_as` for now, but saving this for the future.

File: astropy/coordinates/tests/test_earth.py
Comment (astrojuanlu) [reply astropy/coordinates/tests/test_earth.py:672]: Removed `represent_as` for now, will test it when we finally add it.

File: astropy/coordinates/earth.py
Comment (astrojuanlu) [reply astropy/coordinates/earth.py:875]: Tried this trick https://github.com/astropy/astropy/issues/11063#issuecomment-732212333 let's see if it works!

File: astropy/coordinates/earth.py
Comment (astrojuanlu) [reply astropy/coordinates/earth.py:927]: I ended up removing the docstrings, after all the name should be explicit enough.

However, I did a local `tox -e build_docs` and I didn't see these classes anywhere in the docs, so I added them to `__all__` in `earth.py`.

File: astropy/coordinates/tests/test_earth.py
Comment (astrojuanlu) [reply astropy/coordinates/tests/test_earth.py:442]: I moved these tests to `test_geodetic_representations.py`, but I'm not sure which ones to remove.

File: astropy/coordinates/earth.py
Comment (astrojuanlu) [reply astropy/coordinates/earth.py:927]: I was wrong, removing the docstrings leaves the class documentation empty:

Screenshot_2020-12-01 WGS84GeodeticRepresentation — Astropy v4 3 dev322+g7f34535cc d20201201

Comment (astrojuanlu) [reply astropy/coordinates/earth.py:927]: `@format_doc` proved to be the best solution :+1:

File: astropy/coordinates/tests/test_earth.py
Comment (astrojuanlu) [reply astropy/coordinates/tests/test_earth.py:44]: Done!

File: astropy/coordinates/tests/test_earth.py
Comment (astrojuanlu) [reply astropy/coordinates/tests/test_earth.py:619]: Done!

**Comment by mkbrewer** [comment]
While supporting alternate ellipsoids is fine, you should realize that there is more work involved in doing so. As currently written, `ErfaAstrom.apco()` calls `erfa.apco()` which calculates the observer's position and velocity based on the WGS-84 ellipsoid. Thus geodetic  coordinates based on an alternate ellipsoid will yield incorrect results in any ICRS<->CIRS or ICRS<->AltAz transform.

**Comment by mkbrewer** [comment]
I should note that this relates to task 1) in issue #11072

**Comment by astrojuanlu** [comment]
Warnings are legitimate:

```
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11086/lib/python3.8/site-packages/astropy/coordinates/earth.py:docstring of astropy.coordinates.GRS80GeodeticRepresentation:1: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
writing output... [ 16%] api/astropy.coordinates.SphericalCosLatDifferential .. api/astropy.io.ascii.Latex
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11086/lib/python3.8/site-packages/astropy/coordinates/earth.py:docstring of astropy.coordinates.WGS72GeodeticRepresentation:1: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11086/lib/python3.8/site-packages/astropy/coordinates/earth.py:docstring of astropy.coordinates.WGS84GeodeticRepresentation:1: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
...
[changelog_links] Adding changelog links to "changelog"
<rst_epilog>:14: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
```

fixing

**Comment by astrojuanlu** [comment]
> As currently written, `ErfaAstrom.apco()` calls `erfa.apco()` which calculates the observer's position and velocity based on the WGS-84 ellipsoid. Thus geodetic coordinates based on an alternate ellipsoid will yield incorrect results in any ICRS<->CIRS or ICRS<->AltAz transform.

I don't understand @mkbrewer , isn't `ErfaAstrom.apco()` always converting whatever coordinates to WGS84?

https://github.com/astropy/astropy/blob/9680cd546aa9063758f2c23c836ca79a7c8f1eb1/astropy/coordinates/erfa_astrom.py#L45

Unless I missed anything, the `EarthLocation.to_geodetic` return value didn't change with this PR.

**Comment by astrojuanlu** [comment]
All CI checks passed now :tada:

**Comment by mkbrewer** [comment]
> I don't understand @mkbrewer , isn't ErfaAstrom.apco() always converting whatever coordinates to WGS84?

Oh, you're right. I missed that conversion. Sorry.

File: CHANGES.rst
Comment (mhvk) [inline CHANGES.rst:27]: Should we note what these are good for? E.g., "These are mostly for use inside ``EarthLocation`` but can also be used to convert between geocentric (cartesian) and different geodetic representations directly."

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:304]: Might be clearer to change the unit of `xyz` here instead of `self` at the end - mostly as we can do it in-place.

File: astropy/coordinates/earth.py
Comment (mhvk) [inline astropy/coordinates/earth.py:604]: Can just be `llh.height << self.unit`.

File: astropy/coordinates/representation.py
Comment (mhvk) [inline astropy/coordinates/representation.py:1813]: Maybe keep outside of this PR?

**Review by mhvk — COMMENTED** [review]
This is looking good! It made me realize that really `gc2gd` and `gd2gc` should be made to work with `Quantity` - so I took the liberty of adding a commit that does so.

Also, many of the other comments are a bit nit-picky, so while I was at it, I addressed them. Main thing to perhaps think about is that I removed the use of `Distance` - I find it just too unhandy, not really suitable for a height, and it really is more useful for converting to redshift, etc...

Will push those 2 commits to your branch...

**Comment by mhvk** [comment]
Oops, some failures - will fix (should have tested all of `coordinates`...

**Comment by astrojuanlu** [comment]
Thanks @mhvk ! I can leave the rest to you or address the comments tomorrow, as you prefer

File: astropy/coordinates/earth.py
Comment (mhvk) [reply astropy/coordinates/earth.py:873]: Let's leave this for follow-up...

**Comment by mhvk** [comment]
@astrojuanlu - OK, I also changed the tests now. Plus one thing that you should sign off on: to make `ELLIPSOIDS` a dict with the actual representations as values. I think it works quite nicely, mostly because #11090 already replaced the meta class with `__init_subclass__`, so one can now easily chain things. See 4612ca1.

**Review by mhvk — APPROVED** [review]
Approving, but, as noted, @astrojuanlu, would be best if you approve the changes I made!

File: astropy/coordinates/earth.py
Comment (astrojuanlu) [inline astropy/coordinates/earth.py:879]: Nice trick!

**Review by astrojuanlu — COMMENTED** [review]
Can't approve my own pull request. This looks good to me, thanks @mhvk for the final tweaks!

**Pull request merged**
